### ISSUE TITLE
<meta> is too late?

### ISSUE THREAD

**Issue by annevk** [mystor/isolatewindowproxy#2]
By the time you reach `<meta>`, you've already set up the entire environment. Shouldn't this be an HTTP header?

**Comment by mystor** [comment]
Good point - I'll add the following section to explain my reasoning as to why we initially proposed a <meta> tag, but I want to leave this open, because an HTTP header would certainly be a more elegant solution, and we should consider it.

### ALTERNATIVE: X-IsolateWindowProxy-like HTTP Header

Another option would be to, instead of using a `<meta>` tag, use an HTTP header. The advantage of this is the information that the incoming load should be an `isolatewindowproxy` load would arrive before the environment had already been set up within the old window proxy. The disadvantage is that it would require people to have control over their HTTP headers in order to opt into this form of navigation.

There are forms of navigation, especially with games on gaming websites, or on shared hosting, where developers don't have control over the HTTP headers which are sent as part of the response, but do have control over the HTTP content. One of the desired use cases for this feature would be for games to be able to isolate themselves within their own process. Many game hosting websites don't allow control over the HTTP headers, but adding a <meta> tag would be easy. This is why it was considered initially.

With the <meta> tag, the load has to be canceled and restarted within the new WindowProxy, which wouldn't have to be the case for the HTTP header, which is a serious advantage.

**Comment by lukewagner** [comment]
Additionally, from what I heard from you and bz, HTTP headers are also "too late" in that the overall loading process has already started (at least for a fresh process).  Reliably avoiding the reload is the purpose of the `rel` part of the proposal: this lets you say it before _everything_.

We can also imagine browsers optimizations to avoid the reload even with the `<meta>`:
1. the presence of a contained `<meta>` could be annotated on the HTTP cache entry
2. if the `<meta>` is seen early enough, it seems like the browser could "fix up" the current load so that it can continue.  Or perhaps we do some things more lazily so that seeing the `<meta>` early enough doesn't require any fix-up.  Or, pushing further, we could perhaps enumerate what those conditions are and _require_ the `<meta>` to happen early enough.

**Comment by mystor** [comment]
On the other hand, if we see the HTTP header, we could _theoretically_ redirect the channel over IPC into a different process. It wouldn't be pretty but it potentially could work? Transplanting the entire environment seems much much harder.

**Comment by lukewagner** [comment]
Yeah, for "fixing up", I was talking about same-process use of isolation where my hope was maybe we can delay the creation/proxying of globals until the first event X that cares and thus if we see the `<meta>` before that point, there's no need to restart the load.  For a fresh process, as long as we're still before that first event X, then perhaps the redirection still works just the same as if it was an HTTP header?  I'm just not clear on what all goes into X :)
